### ISSUE TITLE
v2 unintended indentation in `*org-roam*` buffer

### ISSUE THREAD

**Issue by mohkale** [org-roam/org-roam#1586]
### Description

![Screenshot_20210715_183545](https://user-images.githubusercontent.com/23294780/125832219-5ad61482-42a8-487c-b979-54f79820eb00.png)

When `org-adapt-indentation` is `t` org indents sub-headings to match their level but the org-roam buffer seems to just take the first section of the heading including any indentation on lines following the first line. This can be seen in the image above.

#### Steps to Reproduce

1. Load Emacs
2. Run `org-roam--build-cache-async`
3. Run `org-roam-buffer-toggle`
4. Run `org-roam-find-file` and navigate to a file.
5. Create a new sub-heading and insert a link to another file in a paragrap spanning multiple lines.
6. Save all the buffers
7. Run `org-roam-find-file` and go to the other file.

#### Expected Results

Either the backlink contents isn't indented at all or uniformly indented. Preferably the former, perhaps by adapting the code that reads the back-link text to remove leading whitespace up-to the level of the heading.

#### Actual Results

Inconsistent indentation. The first line has no indentation but following ones do.

### Environment

- Org-roam commit: https://github.com/jethrokuan/org-roam/commit/90945e8c03b170731070af2009b89b9687ff2f58
- Emacs: GNU Emacs 28.0.50 (build 3, x86_64-pc-linux-gnu, GTK+ Version 3.24.29, cairo version 1.17.4)
 of 2021-05-24
- Framework: N/A
- Org: Org mode version 9.4.6 (9.4.6-gfc273b @ /home/mohkale/.cache/emacs/straight/build/org/)
- Org-roam: 2.0.0

**Comment by jethrokuan** [comment]
Unrelated, but it looks like you're also affected by `global-page-break-lines-mode`.

**Comment by mohkale** [comment]
@jethrokuan 

Yes, I have it hooked into a few modes although I'm not sure how you realised that from the screenshot :confused: .

```lisp
(use-package page-break-lines
  :straight t
  :hook ((org-mode                    . page-break-lines-mode)
         (help-mode                   . page-break-lines-mode)
         (outline-mode                . page-break-lines-mode)
         (emacs-lisp-compilation-mode . page-break-lines-mode)
         (lisp-interaction-mode       . page-break-lines-mode)
         (emacs-lisp-mode             . page-break-lines-mode))
  :commands (page-break-lines-mode)
  :diminish
  page-break-lines-mode)
```

**Comment by camdez** [comment]
I have a fix for this locally. Making some final decisions and then I'll post up a detailed description of exactly what's happening here (this took many hours of debugging and it's a quirky one) and a PR.

**Comment by real-or-random** [comment]
@camdez Any news on this? :)

**Comment by camdez** [comment]
Sorry for dropping the ball here, got super busy with work and haven't made it back to fully fix the issue. (@real-or-random, thanks for the nudge!)

I don't actually think the fix will be hard (famous last words), it was just hard to figure out, so I'll just share all of my debugging notes and if someone beats me to fixing it and wins all of the internet points, I guess I can live with that. ;)

/cc @purcell Perhaps you would find this interesting as well, as the author of `page-break-lines-mode`.

## Context

-   When using `org-roam-buffer-toggle`, text is sometimes displayed out of order.
-   When using `org-roam-buffer-refresh`, the issue does not occur.

Both codepaths use the exact same rendering code (via`org-roam-buffer-persistent-redisplay`).

## What's Happening

The subtle (but critical) difference between the two calls is that `org-roam-buffer-refresh` is typically run from the window displaying `org-roam-buffer` (it even asserts the right mode), so it has the buffer's window as "the selected window" whereas `org-roam-buffer-toggle` uses `with-current-buffer`, usually with a different selected window.

The selected window shouldn't typically be of consequence, but clearly the value of point is somehow getting mangled (leading to insertion at the wrong location inside calls to `insert-magit-section`).

This lead me to learning about the distinction between a *buffer's* notion of point and a *window's* notion of point.

Per the [Emacs Lisp manual section on Windows and Point](https://www.gnu.org/software/emacs/manual/html_node/elisp/Window-Point.html):

> Each window has its own value of point (see Point), independent of
> the value of point in other windows displaying the same
> buffer. This makes it useful to have multiple windows showing one
> buffer.

Obvious, once you think about it.  Critically it goes on to say:

> Selecting a window sets the value of point in its buffer from the
> window’s value of point. Conversely, deselecting a window sets the
> window’s value of point from that of the buffer. Thus, when you
> switch between windows that display a given buffer, the point value
> for the selected window is in effect in the buffer, while the point
> values for the other windows are stored in those windows.

Just to be clear on that:

                Buffer  Window
                point   point
    Initially:    P1      P2
    Select Win:   P2  <-  P2
    Buf Inserts:  P3      P2
    Deselect Win: P3  ->  P3

(The most important takeaway is that P1 is discarded.)

This lead me to suspect that something was causing the current window to change, leading to the expected value of point being overwritten by the visible window's notion of point.

The culprit is several layers deeper.  To render the contents of the sections with the original syntax highlighting, `org-roam-node-insert-section` calls `org-roam-fontify-like-in-org-mode`, which calls `org-mode` in a temporary buffer containing the text to be formatted.

Any major mode change like this triggers `after-change-major-mode-hook`, which is populated with (autogenerated) functions for any globally-enabled minor modes (defined with `define-global-minor-mode`)&#x2014;and that's where we find the root cause.

In my case&#x2014;and, it sounds like, many others&#x2014;`global-page-break-lines-mode` is the one installing the problematic hook.  There are a couple more layers of intermediate glue code, but ultimately `page-break-lines--update-display-tables` walks the entire window list of the selected frame **calling `with-selected-window` on each one**.  This resets each buffer's notion of point to each window's notion of point, trashing our expected insertion point.

As far as I know, neither `org-roam-mode` nor `page-break-lines-mode` is doing anything inappropriate here per Emacs conventions, but what `page-break-lines-mode` is doing leads to some spooky action at a distance.  I have no idea how many other libraries do something like this.

On the other hand, `org-roam-mode` could be more defensive here and avoid the problem simply by making the `org-roam-buffer`'s window current.

(Could still have issues if multiple windows were displaying the buffer and had different values of point, right?)

**Comment by abgruszecki** [comment]
Great to see there's progress on this! I'm looking forward to the PR :)
